text/0000-variance.md

```diff
@@ -0,0 +1,568 @@
+  In fact, there are only two: `PhantomData` and `PhantomFn`.
+- Integrate the notion of `PhantomData` into other automated compiler
+  analyses, notably OIBIT, that can otherwise be deceived into yielding
+  incorrect results.
+
+# Motivation
+
+## Why variance is good
+
+Today, all type parameters are invariant. This can be problematic
+around lifetimes. A particular common example of where problems
+arise is in the use of `Option`. Here is a simple example. Consider
+this program, which has a struct containing two references:
+
+```
+struct List<'l> {
+    field1: &'l int,
+    field2: &'l int,
+}
+
+fn foo(field1: &int, field2: &int) {
+    let list = List { field1: field1, field2: field2 };
+    ...
+}
+
+fn main() { }
+```
+
+Here the function `foo` takes two references with distinct lifetimes.
+The variable `list` winds up being instantiated with a lifetime that
+is the intersection of the two (presumably, the body of `foo`).  This
+is good.
+
+If we modify this program so that one of those references is optional,
+however, we will find that it gets a compilation error:
+
+```
+struct List<'l> {
+    field1: &'l int,
+    field2: Option<&'l int>,
+}
+
+fn foo(field1: &int, field2: Option<&int>) {
+    let list = List { field1: field1, field2: field2 };
+        // ERROR: Cannot infer an appropriate lifetime
+    ...
+}
+
+fn main() { }
+```
+
+The reason for this is that because `Option` is *invariant* with
+respect to its argument type, it means that the lifetimes of `field1`
+and `field2` must match *exactly*. It is not good enough for them to
+have a common subset. This is not good.
+
+## What variance is
+
+[Variance][v] is a general concept that comes up in all languages that
+combine subtyping and generic types. However, because in Rust all
+subtyping is related to the use of lifetimes parameters, Rust uses
+variance in a very particular way. Basically, variance is a
+determination of when it is ok for lifetimes to be approximated
+(either made bigger or smaller, depending on context).
+
+Let me give a few examples to try and clarify how variance works.
+Consider this simple struct `Context`:
+
+```rust
+struct Context<'data> {
+    data: &'data u32,
+    ...
+}
+```
+
+Here the `Context` struct has one lifetime parameter, `data`, that
+represents the lifetime of some data that it references. Now let's
+imagine that the lifetime of the data is some lifetime we call
+`'x`. If we have a context `cx` of type `Context<'x>`, it is ok to
+(for example) pass `cx` as an argment where a value of type
+`Context<'y>` is required, so long as `'x : 'y` ("`'x` outlives
+`'y`"). That is, it is ok to approximate `'x` as a shorter lifetime
+like `'y`. This makes sense because by changing `'x` to `'y`, we're
+just pretending the data has a shorter lifetime than it actually has,
+which can't do any harm.  Here is an example:
+
+```rust
+fn approx_context<'long,'short>(t: &Context<'long>, data: &'short Data)
```

Comment:
Is this meant to take a reference for the first parameter? As it is, the example is a compile error because `do_something` expects a `Context`, but gets passed a `&Context` in `approx_context`.